network/p2p/p2pnode/internal/cache.go

```diff
@@ -0,0 +1,188 @@
+package internal
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	herocache "github.com/onflow/flow-go/module/mempool/herocache/backdata"
+	"github.com/onflow/flow-go/module/mempool/herocache/backdata/heropool"
+	"github.com/onflow/flow-go/module/mempool/stdmap"
+	"github.com/onflow/flow-go/network"
+)
+
+var (
+	ErrDisallowCacheEntityNotFound = errors.New("disallow list cache entity not found")
+)
+
+// DisallowListCache is the disallow-list cache. It is used to keep track of the disallow-listed peers and the reasons for it.
+type DisallowListCache struct {
+	c *stdmap.Backend
+}
+
+// NewDisallowListCache creates a new disallow-list cache. The cache is backed by a stdmap.Backend.
+// Args:
+// - sizeLimit: the size limit of the cache, i.e., the maximum number of records that the cache can hold, recommended size is 100 * number of authorized nodes.
+// - logger: the logger used by the cache.
+// - collector: the metrics collector used by the cache.
+// Returns:
+// - *DisallowListCache, the created cache.
+func NewDisallowListCache(sizeLimit uint32, logger zerolog.Logger, collector module.HeroCacheMetrics) *DisallowListCache {
+	backData := herocache.NewCache(sizeLimit,
+		herocache.DefaultOversizeFactor,
+		// this cache is supposed to keep the disallow-list causes for the authorized (staked) nodes. Since the number of such nodes is
+		// expected to be small, we do not eject any records from the cache. The cache size must be large enough to hold all
+		// the spam records of the authorized nodes. Also, this cache is keeping at most one record per peer id, so the
+		// size of the cache must be at least the number of authorized nodes.
+		heropool.NoEjection,
+		logger.With().Str("mempool", "disallow-list-records").Logger(),
+		collector)
+
+	return &DisallowListCache{
+		c: stdmap.NewBackend(stdmap.WithBackData(backData)),
+	}
+}
+
+// GetAllDisallowedListCausesFor returns the list of causes for which the given peer is disallow-listed.
+// Args:
+// - peerID: the peer to check.
+// Returns:
+// - the list of causes for which the given peer is disallow-listed. If the peer is not disallow-listed for any reason,
+// an empty list is returned.
+func (d *DisallowListCache) GetAllDisallowedListCausesFor(peerID peer.ID) []network.DisallowListedCause {
+	causes := make([]network.DisallowListedCause, 0)
+	entity, exists := d.c.ByID(makeId(peerID))
```

Comment:
Instead of using the makeId function we should inject the identity provider and use the NodeID from the peer identity instead of computing the ID or change the peerID to nodeID and allow the caller to call funcs from the DisallowListCache with the NodeID. Since makeId returns a flow.Identifier it's better to use the actual NodeID of the peer for storing.

Reply:
Thank you for your input. Generally, the comment is well-articulated. I want to highlight that the current implementation aligns with our objective of achieving a clean delineation between the `Network` and `LibP2PNode` by eliminating the `Middleware` component. This separation is aimed at establishing a distinct division of responsibilities and encapsulation for the `Network` and `LibP2PNode` implementations. As a result, we can create an architecture that is cleaner, more maintainable, and easier to comprehend, debug, and reason about. An integral part of this refactoring process is the application of the [Principle of Least Privilege (PoLP)](https://en.wikipedia.org/wiki/Principle_of_least_privilege), which means that a component should only have access to the information and resources necessary for its legitimate purpose.

In the context of node identifiers, implementing the PoLP can be achieved by differentiating the representation of a node ID between `Network` and `LibP2PNode`. The `LibP2PNode` implementation interacts with both staked and unstaked peers, and it doesn't concern itself with the concept of a Flow-level identifier. Thus, by applying PoLP, we can allow `peer.ID` to serve as the exclusive identifier for a `LibP2PNode` at this level. Conversely, when transitioning to the `Network` layer, which operates within the realm of Flow semantics, the `flow.Identifier` can be adopted as the exclusive identifier.

Given that the code in question is an attribute of the `LibP2PNode`, it's prudent to continue using the nodes' `peer.ID` for querying. This approach also yields additional benefits:
1. The same cache can cater to both staked and unstaked nodes, as well as public and private networks, solely by leveraging their `peer.ID`.
2. Since `LibP2PNode` does not require an identity provider (or translator) for its operations, retaining the current code without adding an identity provider minimizes dependencies between the foundational `LibP2PNode` and the more advanced Flow constructs. This enhances coherence and simplicity within the structure.